# Grundbegriffe 

Beantworte und diskutiere die folgenden Fragen konkret für den in Einheit 1 heruntergeladenen San Francisco Library Usage Datensatz. Notiere Deine Ergebnisse in Stichpunkte.

In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv("/home/katja/2024-2025-Data_Lirbrarian_Katja_Gödde/Modul_3/data/Library_Usage.csv", dtype={"Within San Francisco County": str}, na_values=['Null', 'NA', ''], low_memory=False)  

In [4]:
df

Patron Type Definition  Total Checkouts  Total Renewals  \
0                      Senior                5               0   
1                       Adult                0               0   
2                       Adult                0               0   
3                     Welcome                1               1   
4                      Senior                0               0   
...                       ...              ...             ...   
450354    Digital Access Card                0               0   
450355    Digital Access Card                0               0   
450356    Digital Access Card                0               0   
450357    Digital Access Card                0               0   
450358    Digital Access Card                0               0   

                Age Range Home Library Definition Circulation Active Month  \
0       75 years and over                    Main                      Nov   
1          45 to 54 years                    Main                      Jul   
2          55 to 59 years        Western Addition                      Mar   
3          20 to 24 years                Richmond                      Aug   
4          65 to 74 years                  Sunset                      Mar   
...                   ...                     ...                      ...   
450354     35 to 44 years               Ingleside                      NaN   
450355     45 to 54 years               Ingleside                      Aug   
450356     25 to 34 years               Ingleside                      Aug   
450357     35 to 44 years               Ingleside                      Apr   
450358     65 to 74 years                 Bayview                      Apr   

        Circulation Active Year Notice Preference Definition  \
0                        2022.0                        Email   
1                        2023.0                        Email   
2                        2024.0                        Email   
3                        2022.0                        Email   
4                        2024.0                        Print   
...                         ...                          ...   
450354                      NaN                        Email   
450355                   2022.0                        Print   
450356                   2022.0                        Print   
450357                   2022.0                        Print   
450358                   2023.0                        Email   

        Provided Email Address  Year Patron Registered  \
0                         True                    2015   
1                         True                    2019   
2                         True                    2022   
3                         True                    2022   
4                        False                    2023   
...                        ...                     ...   
450354                   False                    2023   
450355                   False                    2022   
450356                   False                    2022   
450357                   False                    2022   
450358                   False                    2023   

       Within San Francisco County  
0                            False  
1                            False  
2                            False  
3                            False  
4                            False  
...                            ...  
450354                         NaN  
450355                         NaN  
450356                         NaN  
450357                         NaN  
450358                         NaN  

[450359 rows x 11 columns]

Wie viele Merkmale besitzt der Datensatz?
Der Datensatz besitzt 11 Merkmale, die hier durch die Spaltennamen repräsentiert werden.

Wie groß ist die Stichprobengröße des Datensatzes? 450359 (= die Anzahl der Zeilen)

Wer oder was sind die Merkmalsträger? Die Merkmalsträger sind die Bibliotheksnutzer*innen an denen die Ausprägungen gemessen werden.

Von wann bis wann wurden die Daten erhoben? Vom 02. Dezember 2016 bis zum 25. März 2024 (einsehbar in den Metadaten unter https://data.sfgov.org/Culture-and-Recreation/Library-Usage/qzz6-2jup/about_data)

Wie lässt sich die Grundgesamtheit beschreiben? Handelt es sich um eine Vollerhebung? Es handelt sich nicht um eine Vollerhebung, da nicht alle Kunden, sondern nur aktive Kunden mit einbezogen wurden. Aktive Kunden sind Kunden mit noch nicht abgelaufenen Bibliotheksausweisen und Kunden, die innerhalb der letzten drei Jahre eine Ausleihtätigkeit ausgeübt haben.

Welche Merkmale sind stetig? Welche diskret?
- Patron Type Definition: Diskret, da nur abzählbar viele Werte angenommen werden können.
- Total Checkouts: Diskret, da nur abzählbar viele Werte angenommen werden können.
- Total Reneweals: Diskret, da nur abzählbar viele Werte angenommen werden können.
- Age Range: Diskret, da es sich um abzählbar viele Gruppierungen von Altersklassen handelt. Wenn es keine Gruppierung gäbe, wäre das Merkmal dagegen stetig.
- Home Library Definition: Diskret, da nur abzählbar viele Werte angenommen werden können.
- Circulation Active Month: Diskret, da nur abzählbar viele Werte angenommen werden können.
- Circulation Active Year: Diskret, da nur abzählbar viele Werte angenommen werden können.
- Notice Preference Definition: Diskret, da nur abzählbar viele Werte angenommen werden können.
- Provided Email Address: Diskret, da nur abzählbar viele Werte angenommen werden können.
- Year Patron Registered: Diskret, da nur abzählbar viele Werte angenommen werden können.
- Within San Francisco County: Diskret, da nur abzählbar viele Werte angenommen werden können.

Welchem Skalenniveau entsprechen die einzelnen Merkmale (Nominal-, Ordinal- oder Metrische Skala)?
- Patron Type Definition: Nominalskala
- Total Checkouts: Metrische Skala
- Total Reneweals: Metrische Skala
- Age Range: Ordinalskala
- Home Library Definition: Nominalskala
- Circulation Active Month: Ordinalskala
- Circulation Active Year: Metrische Skala
- Notice Preference Definition: Nominalskala
- Provided Email Address: Nominalskala
- Year Patron Registered: Metrische Skala
- Within San Francisco County: Nominalskala

Enthält der Datensatz fehlende Werte? Ja, in den Spalten Age Range, Home Library Definition, Circulation Active Month, Circulation Active Year, Notice Preference Definition und Within San Francisco County

In [13]:
fehlende_werte = df.isnull().sum()
print(fehlende_werte)

Patron Type Definition              0
Total Checkouts                     0
Total Renewals                      0
Age Range                         510
Home Library Definition            30
Circulation Active Month        40317
Circulation Active Year         40317
Notice Preference Definition     3315
Provided Email Address              0
Year Patron Registered              0
Within San Francisco County       221
dtype: int64


Handelt es sich um Querschnitts-, Längsschnitss- oder Paneldaten? Es handelt sich um Paneldaten, da Beobachtungen von denselben Einheiten (Bibliotheksnutzer*innen) zu verschiedenen Zeitpunkten erhoben werden. 

# 2.5 Fallstudie: Feature Engineering

In [16]:
df.columns

Index(['Patron Type Definition', 'Total Checkouts', 'Total Renewals',
       'Age Range', 'Home Library Definition', 'Circulation Active Month',
       'Circulation Active Year', 'Notice Preference Definition',
       'Provided Email Address', 'Year Patron Registered',
       'Within San Francisco County'],
      dtype='object')

In [17]:
df['Total Checkouts']

0         5
1         0
2         0
3         1
4         0
         ..
450354    0
450355    0
450356    0
450357    0
450358    0
Name: Total Checkouts, Length: 450359, dtype: int64

In [18]:
df[['Total Checkouts', 'Total Renewals']]

Total Checkouts  Total Renewals
0                     5               0
1                     0               0
2                     0               0
3                     1               1
4                     0               0
...                 ...             ...
450354                0               0
450355                0               0
450356                0               0
450357                0               0
450358                0               0

[450359 rows x 2 columns]

In [19]:
df['is_adult'] = df['Patron Type Definition'] == 'Adult'
df['log_renewals'] = np.log(df['Total Renewals'] + 1)

In [20]:
df['is_adult']

0         False
1          True
2          True
3         False
4         False
          ...  
450354    False
450355    False
450356    False
450357    False
450358    False
Name: is_adult, Length: 450359, dtype: bool

In [21]:
df['log_renewals']

0         0.000000
1         0.000000
2         0.000000
3         0.693147
4         0.000000
            ...   
450354    0.000000
450355    0.000000
450356    0.000000
450357    0.000000
450358    0.000000
Name: log_renewals, Length: 450359, dtype: float64

In [22]:
df

Patron Type Definition  Total Checkouts  Total Renewals  \
0                      Senior                5               0   
1                       Adult                0               0   
2                       Adult                0               0   
3                     Welcome                1               1   
4                      Senior                0               0   
...                       ...              ...             ...   
450354    Digital Access Card                0               0   
450355    Digital Access Card                0               0   
450356    Digital Access Card                0               0   
450357    Digital Access Card                0               0   
450358    Digital Access Card                0               0   

                Age Range Home Library Definition Circulation Active Month  \
0       75 years and over                    Main                      Nov   
1          45 to 54 years                    Main                      Jul   
2          55 to 59 years        Western Addition                      Mar   
3          20 to 24 years                Richmond                      Aug   
4          65 to 74 years                  Sunset                      Mar   
...                   ...                     ...                      ...   
450354     35 to 44 years               Ingleside                      NaN   
450355     45 to 54 years               Ingleside                      Aug   
450356     25 to 34 years               Ingleside                      Aug   
450357     35 to 44 years               Ingleside                      Apr   
450358     65 to 74 years                 Bayview                      Apr   

        Circulation Active Year Notice Preference Definition  \
0                        2022.0                        Email   
1                        2023.0                        Email   
2                        2024.0                        Email   
3                        2022.0                        Email   
4                        2024.0                        Print   
...                         ...                          ...   
450354                      NaN                        Email   
450355                   2022.0                        Print   
450356                   2022.0                        Print   
450357                   2022.0                        Print   
450358                   2023.0                        Email   

        Provided Email Address  Year Patron Registered  \
0                         True                    2015   
1                         True                    2019   
2                         True                    2022   
3                         True                    2022   
4                        False                    2023   
...                        ...                     ...   
450354                   False                    2023   
450355                   False                    2022   
450356                   False                    2022   
450357                   False                    2022   
450358                   False                    2023   

       Within San Francisco County  is_adult  log_renewals  
0                            False     False      0.000000  
1                            False      True      0.000000  
2                            False      True      0.000000  
3                            False     False      0.693147  
4                            False     False      0.000000  
...                            ...       ...           ...  
450354                         NaN     False      0.000000  
450355                         NaN     False      0.000000  
450356                         NaN     False      0.000000  
450357                         NaN     False      0.000000  
450358                         NaN     False      0.000000  

[450359 rows x 13 columns]

In [23]:
df['Membership Duration'] = ('Circulation Active Year' - 'Year Patron Registered')*12 + 'Circulation Active Month'

TypeError: unsupported operand type(s) for -: 'str' and 'str'

In [33]:
pd.to_numeric(
  df['Circulation Active Year'], errors='coerce'
)

0         2022.0
1         2023.0
2         2024.0
3         2022.0
4         2024.0
           ...  
450354       NaN
450355    2022.0
450356    2022.0
450357    2022.0
450358    2023.0
Name: Circulation Active Year, Length: 450359, dtype: float64

In [35]:
pd.to_datetime(
    df['Circulation Active Month'],
    errors='coerce',
    format="%b"
)


0        1900-11-01
1        1900-07-01
2        1900-03-01
3        1900-08-01
4        1900-03-01
            ...    
450354          NaT
450355   1900-08-01
450356   1900-08-01
450357   1900-04-01
450358   1900-04-01
Name: Circulation Active Month, Length: 450359, dtype: datetime64[ns]

In [37]:
df['Circulation Active Month'].dt.month

AttributeError: Can only use .dt accessor with datetimelike values

In [39]:
df['Circulation Active Month'].head() 

0    Nov
1    Jul
2    Mar
3    Aug
4    Mar
Name: Circulation Active Month, dtype: object

In [41]:
df['Circulation Active Month'].unique()

array(['Nov', 'Jul', 'Mar', 'Aug', 'Apr', 'Feb', 'Jan', nan, 'May', 'Oct',
       'Sep', 'Jun', 'Dec'], dtype=object)

In [43]:
df['Membership Duration'] = ('Circulation Active Year' - 'Year Patron Registered')*12 + 'Circulation Active Month'


TypeError: unsupported operand type(s) for -: 'str' and 'str'

In [45]:
df['Membership Duration'].fillna(0)

KeyError: 'Membership Duration'

# Aufgabe 2.6

Beispiele nachvollziehen

In [47]:
df.loc[df['Total Checkouts'] > 10000]

Patron Type Definition  Total Checkouts  Total Renewals  \
12694                  Senior            12233            5416   
12925                  Senior            10428            1079   
54102                  Senior            17663              56   
54414                  Senior            12675            1422   
59475                  Senior            11252            2628   
...                       ...              ...             ...   
293994                  Adult            39743             659   
294100                  Adult            10454            8663   
306622                  Adult            10065            3897   
312053                  Adult            14825            3915   
328352                  Adult            12122             122   

                Age Range Home Library Definition Circulation Active Month  \
12694   75 years and over                  Merced                      Mar   
12925   75 years and over                  Ortega                      Mar   
54102      65 to 74 years                  Marina                      Mar   
54414      65 to 74 years               Chinatown                      Mar   
59475      65 to 74 years                    Main                      Mar   
...                   ...                     ...                      ...   
293994     60 to 64 years                  Ortega                      Mar   
294100     35 to 44 years                  Ortega                      Mar   
306622     60 to 64 years      Golden Gate Valley                      Mar   
312053     25 to 34 years                    Park                      Mar   
328352     55 to 59 years                 Bayview                      Mar   

        Circulation Active Year Notice Preference Definition  \
12694                    2024.0                        Email   
12925                    2024.0                        Email   
54102                    2024.0                        Email   
54414                    2024.0                        Email   
59475                    2024.0                        Phone   
...                         ...                          ...   
293994                   2024.0                        Email   
294100                   2024.0                        Email   
306622                   2024.0                        Email   
312053                   2024.0                        Email   
328352                   2024.0                        Email   

        Provided Email Address  Year Patron Registered  \
12694                     True                    2003   
12925                     True                    2003   
54102                     True                    2008   
54414                     True                    2003   
59475                    False                    2013   
...                        ...                     ...   
293994                    True                    2003   
294100                    True                    2003   
306622                    True                    2003   
312053                    True                    2003   
328352                    True                    2003   

       Within San Francisco County  is_adult  log_renewals  
12694                        False     False      8.597297  
12925                        False     False      6.984716  
54102                        False     False      4.043051  
54414                        False     False      7.260523  
59475                         True     False      7.874359  
...                            ...       ...           ...  
293994                        True      True      6.492240  
294100                        True      True      9.066932  
306622                        True      True      8.268219  
312053                        True      True      8.272826  
328352                        True      True      4.812184  

[89 rows x 13 columns]

In [49]:
row_filter = df['Total Checkouts'] > 10000
df.loc[row_filter]

Patron Type Definition  Total Checkouts  Total Renewals  \
12694                  Senior            12233            5416   
12925                  Senior            10428            1079   
54102                  Senior            17663              56   
54414                  Senior            12675            1422   
59475                  Senior            11252            2628   
...                       ...              ...             ...   
293994                  Adult            39743             659   
294100                  Adult            10454            8663   
306622                  Adult            10065            3897   
312053                  Adult            14825            3915   
328352                  Adult            12122             122   

                Age Range Home Library Definition Circulation Active Month  \
12694   75 years and over                  Merced                      Mar   
12925   75 years and over                  Ortega                      Mar   
54102      65 to 74 years                  Marina                      Mar   
54414      65 to 74 years               Chinatown                      Mar   
59475      65 to 74 years                    Main                      Mar   
...                   ...                     ...                      ...   
293994     60 to 64 years                  Ortega                      Mar   
294100     35 to 44 years                  Ortega                      Mar   
306622     60 to 64 years      Golden Gate Valley                      Mar   
312053     25 to 34 years                    Park                      Mar   
328352     55 to 59 years                 Bayview                      Mar   

        Circulation Active Year Notice Preference Definition  \
12694                    2024.0                        Email   
12925                    2024.0                        Email   
54102                    2024.0                        Email   
54414                    2024.0                        Email   
59475                    2024.0                        Phone   
...                         ...                          ...   
293994                   2024.0                        Email   
294100                   2024.0                        Email   
306622                   2024.0                        Email   
312053                   2024.0                        Email   
328352                   2024.0                        Email   

        Provided Email Address  Year Patron Registered  \
12694                     True                    2003   
12925                     True                    2003   
54102                     True                    2008   
54414                     True                    2003   
59475                    False                    2013   
...                        ...                     ...   
293994                    True                    2003   
294100                    True                    2003   
306622                    True                    2003   
312053                    True                    2003   
328352                    True                    2003   

       Within San Francisco County  is_adult  log_renewals  
12694                        False     False      8.597297  
12925                        False     False      6.984716  
54102                        False     False      4.043051  
54414                        False     False      7.260523  
59475                         True     False      7.874359  
...                            ...       ...           ...  
293994                        True      True      6.492240  
294100                        True      True      9.066932  
306622                        True      True      8.268219  
312053                        True      True      8.272826  
328352                        True      True      4.812184  

[89 rows x 13 columns]

In [53]:
row_filter = (df['Patron Type Definition'] == 'Senior') & (df['Notice Preference Definition'] == 'Email')
df.loc[row_filter]
# Es werden alle Zeilen ausgegeben, in denen das Age Range der Gruppe der 65-74 jährigen oder der Gruppe der über 75-jährigen entspricht 
# und die Email als Preference ausgewählt haben.

Patron Type Definition  Total Checkouts  Total Renewals  \
0                      Senior                5               0   
5                      Senior                0               0   
27                     Senior                0               0   
28                     Senior                0               0   
29                     Senior                0               0   
...                       ...              ...             ...   
450101                 Senior                0               0   
450102                 Senior                0               0   
450142                 Senior                0               0   
450143                 Senior                0               0   
450144                 Senior                1               0   

                Age Range Home Library Definition Circulation Active Month  \
0       75 years and over                    Main                      Nov   
5       75 years and over                    Main                      Apr   
27      75 years and over                    Main                      Oct   
28         65 to 74 years                    Main                      Mar   
29         65 to 74 years                    Main                      Aug   
...                   ...                     ...                      ...   
450101  75 years and over                    Main                      Mar   
450102     65 to 74 years              Noe Valley                      Mar   
450142     65 to 74 years                    Main                      Nov   
450143     65 to 74 years                    Main                      Mar   
450144  75 years and over                    Main                      Jan   

        Circulation Active Year Notice Preference Definition  \
0                        2022.0                        Email   
5                        2021.0                        Email   
27                       2023.0                        Email   
28                       2024.0                        Email   
29                       2023.0                        Email   
...                         ...                          ...   
450101                   2024.0                        Email   
450102                   2024.0                        Email   
450142                   2022.0                        Email   
450143                   2024.0                        Email   
450144                   2024.0                        Email   

        Provided Email Address  Year Patron Registered  \
0                         True                    2015   
5                         True                    2009   
27                        True                    2016   
28                        True                    2016   
29                        True                    2020   
...                        ...                     ...   
450101                    True                    2013   
450102                    True                    2019   
450142                    True                    2021   
450143                    True                    2019   
450144                    True                    2015   

       Within San Francisco County  is_adult  log_renewals  
0                            False     False           0.0  
5                            False     False           0.0  
27                           False     False           0.0  
28                           False     False           0.0  
29                           False     False           0.0  
...                            ...       ...           ...  
450101                       False     False           0.0  
450102                       False     False           0.0  
450142                         NaN     False           0.0  
450143                         NaN     False           0.0  
450144                         NaN     False           0.0  

[40940 rows x 13 columns]

In [55]:
filter1 = (df['Total Checkouts'] >= 20) & (df['Total Checkouts'] <= 80)
filter2 =  df['Total Checkouts'].between(20, 80)
all(filter1 == filter2)

True

In [57]:
filter1

0         False
1         False
2         False
3         False
4         False
          ...  
450354    False
450355    False
450356    False
450357    False
450358    False
Name: Total Checkouts, Length: 450359, dtype: bool

In [59]:
filter2

0         False
1         False
2         False
3         False
4         False
          ...  
450354    False
450355    False
450356    False
450357    False
450358    False
Name: Total Checkouts, Length: 450359, dtype: bool

Bearbeiten der Aufgabe 2.6

Aufgabe 1: Filtere den Datensatz nach Kindern unter 10 Jahren. Wie viele Einträge erhältst Du? 

In [66]:
df

Patron Type Definition  Total Checkouts  Total Renewals  \
0                      Senior                5               0   
1                       Adult                0               0   
2                       Adult                0               0   
3                     Welcome                1               1   
4                      Senior                0               0   
...                       ...              ...             ...   
450354    Digital Access Card                0               0   
450355    Digital Access Card                0               0   
450356    Digital Access Card                0               0   
450357    Digital Access Card                0               0   
450358    Digital Access Card                0               0   

                Age Range Home Library Definition Circulation Active Month  \
0       75 years and over                    Main                      Nov   
1          45 to 54 years                    Main                      Jul   
2          55 to 59 years        Western Addition                      Mar   
3          20 to 24 years                Richmond                      Aug   
4          65 to 74 years                  Sunset                      Mar   
...                   ...                     ...                      ...   
450354     35 to 44 years               Ingleside                      NaN   
450355     45 to 54 years               Ingleside                      Aug   
450356     25 to 34 years               Ingleside                      Aug   
450357     35 to 44 years               Ingleside                      Apr   
450358     65 to 74 years                 Bayview                      Apr   

        Circulation Active Year Notice Preference Definition  \
0                        2022.0                        Email   
1                        2023.0                        Email   
2                        2024.0                        Email   
3                        2022.0                        Email   
4                        2024.0                        Print   
...                         ...                          ...   
450354                      NaN                        Email   
450355                   2022.0                        Print   
450356                   2022.0                        Print   
450357                   2022.0                        Print   
450358                   2023.0                        Email   

        Provided Email Address  Year Patron Registered  \
0                         True                    2015   
1                         True                    2019   
2                         True                    2022   
3                         True                    2022   
4                        False                    2023   
...                        ...                     ...   
450354                   False                    2023   
450355                   False                    2022   
450356                   False                    2022   
450357                   False                    2022   
450358                   False                    2023   

       Within San Francisco County  is_adult  log_renewals  
0                            False     False      0.000000  
1                            False      True      0.000000  
2                            False      True      0.000000  
3                            False     False      0.693147  
4                            False     False      0.000000  
...                            ...       ...           ...  
450354                         NaN     False      0.000000  
450355                         NaN     False      0.000000  
450356                         NaN     False      0.000000  
450357                         NaN     False      0.000000  
450358                         NaN     False      0.000000  

[450359 rows x 13 columns]

In [70]:
df['Age Range'].unique()

array(['75 years and over', '45 to 54 years', '55 to 59 years',
       '20 to 24 years', '65 to 74 years', '10 to 19 years',
       '25 to 34 years', '60 to 64 years', '35 to 44 years',
       '0 to 9 years', nan], dtype=object)

In [72]:
df.loc[df['Age Range'] == '0 to 9 years']

Patron Type Definition  Total Checkouts  Total Renewals     Age Range  \
14               Teacher Card                0               0  0 to 9 years   
1067                    Adult                0               0  0 to 9 years   
1487                 Juvenile                0               0  0 to 9 years   
1488                 Juvenile                0               0  0 to 9 years   
1491                 Juvenile                0               0  0 to 9 years   
...                       ...              ...             ...           ...   
450241               Juvenile                0               0  0 to 9 years   
450244               Juvenile                2               2  0 to 9 years   
450245               Juvenile                0               0  0 to 9 years   
450246               Juvenile                2               2  0 to 9 years   
450247               Juvenile               11              15  0 to 9 years   

       Home Library Definition Circulation Active Month  \
14                        Main                      NaN   
1067                  Richmond                      Dec   
1487                      Main                      NaN   
1488                      Main                      NaN   
1491                      Main                      Jun   
...                        ...                      ...   
450241          Bernal Heights                      NaN   
450244                 Bayview                      Oct   
450245                 Bayview                      NaN   
450246                 Bayview                      Oct   
450247                 Bayview                      Mar   

        Circulation Active Year Notice Preference Definition  \
14                          NaN                        Email   
1067                     2022.0                        Phone   
1487                        NaN                        Email   
1488                        NaN                        Email   
1491                     2023.0                        Email   
...                         ...                          ...   
450241                      NaN                        Email   
450244                   2023.0                        Email   
450245                      NaN                        Email   
450246                   2023.0                        Email   
450247                   2022.0                        Email   

        Provided Email Address  Year Patron Registered  \
14                        True                    2023   
1067                     False                    2022   
1487                      True                    2020   
1488                      True                    2020   
1491                      True                    2017   
...                        ...                     ...   
450241                    True                    2022   
450244                   False                    2023   
450245                   False                    2023   
450246                   False                    2023   
450247                    True                    2022   

       Within San Francisco County  is_adult  log_renewals  
14                           False     False      0.000000  
1067                         False      True      0.000000  
1487                         False     False      0.000000  
1488                         False     False      0.000000  
1491                         False     False      0.000000  
...                            ...       ...           ...  
450241                         NaN     False      0.000000  
450244                         NaN     False      1.098612  
450245                         NaN     False      0.000000  
450246                         NaN     False      1.098612  
450247                         NaN     False      2.772589  

[33180 rows x 13 columns]

Lösung: Es gibt 33180 Personen unter 10 Jahren.

Aufgabe 2: Gibt es Personen mit mehr als 20000 Ausleihen? 

In [81]:
df.loc[df['Total Checkouts'] > 20000]

Patron Type Definition  Total Checkouts  Total Renewals  \
60282                  Senior            20561            1021   
60622                  Senior            23134             617   
60994                  Senior            21890              16   
179460                  Adult            35176              62   
234329                  Adult            23345            6458   
234381                  Adult            24386             387   
293994                  Adult            39743             659   

                Age Range Home Library Definition Circulation Active Month  \
60282      65 to 74 years                    Main                      Mar   
60622      65 to 74 years                    Main                      Mar   
60994   75 years and over                    Main                      Jan   
179460     25 to 34 years                  Sunset                      Mar   
234329     55 to 59 years               Chinatown                      Mar   
234381     45 to 54 years               Chinatown                      Mar   
293994     60 to 64 years                  Ortega                      Mar   

        Circulation Active Year Notice Preference Definition  \
60282                    2024.0                        Phone   
60622                    2024.0                        Print   
60994                    2023.0                        Email   
179460                   2024.0                        Email   
234329                   2024.0                        Email   
234381                   2024.0                        Phone   
293994                   2024.0                        Email   

        Provided Email Address  Year Patron Registered  \
60282                    False                    2004   
60622                    False                    2003   
60994                     True                    2005   
179460                    True                    2010   
234329                    True                    2003   
234381                   False                    2003   
293994                    True                    2003   

       Within San Francisco County  is_adult  log_renewals  
60282                         True     False      6.929517  
60622                         True     False      6.426488  
60994                         True     False      2.833213  
179460                        True      True      4.143135  
234329                        True      True      8.773230  
234381                        True      True      5.961005  
293994                        True      True      6.492240

Es gibt 7 Personen mit mehr als 20000 Ausleihen

Aufgabe 3: Wie viele Personen stammen aus dem Stadtteil (Richmond)?

In [88]:
df.loc[df['Home Library Definition'] == 'Richmond']

Patron Type Definition  Total Checkouts  Total Renewals  \
3                     Welcome                1               1   
228                    Senior                0               0   
229                    Senior                0               0   
230                    Senior                7              19   
231                    Senior                0               0   
...                       ...              ...             ...   
450159                  Adult                0               0   
450198               Juvenile                0               0   
450255                   Teen                1               0   
450256                   Teen             1288             109   
450291    Digital Access Card                0               0   

                Age Range Home Library Definition Circulation Active Month  \
3          20 to 24 years                Richmond                      Aug   
228        65 to 74 years                Richmond                      Feb   
229     75 years and over                Richmond                      Feb   
230        65 to 74 years                Richmond                      Mar   
231        65 to 74 years                Richmond                      Jan   
...                   ...                     ...                      ...   
450159     25 to 34 years                Richmond                      NaN   
450198     10 to 19 years                Richmond                      Jul   
450255     20 to 24 years                Richmond                      Feb   
450256     10 to 19 years                Richmond                      Aug   
450291     35 to 44 years                Richmond                      Apr   

        Circulation Active Year Notice Preference Definition  \
3                        2022.0                        Email   
228                      2024.0                        Email   
229                      2024.0                        Email   
230                      2024.0                        Email   
231                      2024.0                        Email   
...                         ...                          ...   
450159                      NaN                        Email   
450198                   2023.0                        Email   
450255                   2024.0                        Email   
450256                   2023.0                        Email   
450291                   2023.0                        Email   

        Provided Email Address  Year Patron Registered  \
3                         True                    2022   
228                       True                    2016   
229                       True                    2016   
230                       True                    2023   
231                       True                    2019   
...                        ...                     ...   
450159                    True                    2023   
450198                   False                    2023   
450255                   False                    2023   
450256                    True                    2009   
450291                   False                    2023   

       Within San Francisco County  is_adult  log_renewals  
3                            False     False      0.693147  
228                          False     False      0.000000  
229                          False     False      0.000000  
230                          False     False      2.995732  
231                          False     False      0.000000  
...                            ...       ...           ...  
450159                         NaN      True      0.000000  
450198                         NaN     False      0.000000  
450255                         NaN     False      0.000000  
450256                         NaN     False      4.700480  
450291                         NaN     False      0.000000  

[21433 rows x 13 columns]

21433 Personen stammen aus dem Stadtteil Richmond.

Aufgabe 4: Wie viele Prozent der Beobachtungen haben eine Membership Duration von Null Monaten?

Hier hab ich keine Lösung, weil die Bearbeitung der Aufgabe 2.5 in der die Membership Duration berechnet wurde bei mir nicht funktioniert hat.

# Aufgabe 2.9

In [93]:
df.head(100)

Patron Type Definition  Total Checkouts  Total Renewals          Age Range  \
0                  Senior                5               0  75 years and over   
1                   Adult                0               0     45 to 54 years   
2                   Adult                0               0     55 to 59 years   
3                 Welcome                1               1     20 to 24 years   
4                  Senior                0               0     65 to 74 years   
..                    ...              ...             ...                ...   
95                 Senior               81              37  75 years and over   
96                 Senior                0               0     65 to 74 years   
97                 Senior                0               0     65 to 74 years   
98                 Senior                0               0     65 to 74 years   
99                 Senior                1               0  75 years and over   

   Home Library Definition Circulation Active Month  Circulation Active Year  \
0                     Main                      Nov                   2022.0   
1                     Main                      Jul                   2023.0   
2         Western Addition                      Mar                   2024.0   
3                 Richmond                      Aug                   2022.0   
4                   Sunset                      Mar                   2024.0   
..                     ...                      ...                      ...   
95                    Main                      Jun                   2023.0   
96                    Main                      Oct                   2023.0   
97                    Main                      Mar                   2024.0   
98                    Main                      Mar                   2024.0   
99                    Main                      Mar                   2020.0   

   Notice Preference Definition  Provided Email Address  \
0                         Email                    True   
1                         Email                    True   
2                         Email                    True   
3                         Email                    True   
4                         Print                   False   
..                          ...                     ...   
95                        Email                    True   
96                          NaN                   False   
97                        Email                    True   
98                        Email                    True   
99                        Email                    True   

    Year Patron Registered Within San Francisco County  is_adult  log_renewals  
0                     2015                       False     False      0.000000  
1                     2019                       False      True      0.000000  
2                     2022                       False      True      0.000000  
3                     2022                       False     False      0.693147  
4                     2023                       False     False      0.000000  
..                     ...                         ...       ...           ...  
95                    2003                       False     False      3.637586  
96                    2022                       False     False      0.000000  
97                    2019                       False     False      0.000000  
98                    2023                       False     False      0.000000  
99                    2015                       False     False      0.000000  

[100 rows x 13 columns]

In [95]:
df.describe(include='all')

Patron Type Definition  Total Checkouts  Total Renewals  \
count                  450359    450359.000000   450359.000000   
unique                     17              NaN             NaN   
top                     Adult              NaN             NaN   
freq                   284671              NaN             NaN   
mean                      NaN       155.218053       78.968929   
std                       NaN       526.242567      305.915233   
min                       NaN         0.000000        0.000000   
25%                       NaN         0.000000        0.000000   
50%                       NaN         7.000000        3.000000   
75%                       NaN        74.000000       36.000000   
max                       NaN     39743.000000    12730.000000   

             Age Range Home Library Definition Circulation Active Month  \
count           449849                  450329                   410042   
unique              10                      29                       12   
top     25 to 34 years                    Main                      Mar   
freq             95278                  145932                   150957   
mean               NaN                     NaN                      NaN   
std                NaN                     NaN                      NaN   
min                NaN                     NaN                      NaN   
25%                NaN                     NaN                      NaN   
50%                NaN                     NaN                      NaN   
75%                NaN                     NaN                      NaN   
max                NaN                     NaN                      NaN   

        Circulation Active Year Notice Preference Definition  \
count             410042.000000                       447044   
unique                      NaN                            3   
top                         NaN                        Email   
freq                        NaN                       409071   
mean                2022.934855                          NaN   
std                    1.419927                          NaN   
min                 2004.000000                          NaN   
25%                 2022.000000                          NaN   
50%                 2023.000000                          NaN   
75%                 2024.000000                          NaN   
max                 2024.000000                          NaN   

       Provided Email Address  Year Patron Registered  \
count                  450359           450359.000000   
unique                      2                     NaN   
top                      True                     NaN   
freq                   408532                     NaN   
mean                      NaN             2016.094902   
std                       NaN                6.278639   
min                       NaN             2003.000000   
25%                       NaN             2012.000000   
50%                       NaN             2018.000000   
75%                       NaN             2021.000000   
max                       NaN             2023.000000   

       Within San Francisco County is_adult   log_renewals  
count                       450138   450359  450359.000000  
unique                           2        2            NaN  
top                           True     True            NaN  
freq                        377329   284671            NaN  
mean                           NaN      NaN       1.988588  
std                            NaN      NaN       2.121580  
min                            NaN      NaN       0.000000  
25%                            NaN      NaN       0.000000  
50%                            NaN      NaN       1.386294  
75%                            NaN      NaN       3.610918  
max                            NaN      NaN       9.451795

In [99]:
print(df['Total Renewals'].min())

0


In [103]:
df['Total Renewals'].sum()

35564368

In [105]:
df['Total Renewals'].between(100, 200).sum()

25835

In [109]:
df.columns

Index(['Patron Type Definition', 'Total Checkouts', 'Total Renewals',
       'Age Range', 'Home Library Definition', 'Circulation Active Month',
       'Circulation Active Year', 'Notice Preference Definition',
       'Provided Email Address', 'Year Patron Registered',
       'Within San Francisco County', 'is_adult', 'log_renewals'],
      dtype='object')

In [111]:
df.dtypes

Patron Type Definition           object
Total Checkouts                   int64
Total Renewals                    int64
Age Range                        object
Home Library Definition          object
Circulation Active Month         object
Circulation Active Year         float64
Notice Preference Definition     object
Provided Email Address             bool
Year Patron Registered            int64
Within San Francisco County      object
is_adult                           bool
log_renewals                    float64
dtype: object

In [113]:
df.shape

(450359, 13)